_posts/2016-12-27-angular-2-advance-testing-with-custom-matchers.md

```diff
@@ -0,0 +1,650 @@
+
+
+###### Configuring the TestBed, Component Shell, and Helpers
+
+Very similar to the TestBed sample shown above, I will configure a testing module but I will 
+not *import* an external test component. My test component `TestLayoutComponent` is itself defined 
+within our test (*.spec) file.
+
+> Using an internal test component enables each `<directive>.spec.ts` to define and use its own 
+custom test component with custom properties.
+
+Here is the initial configuration:
+
+{% highlight js %}
+{% raw %}
+import {Component, OnInit} from '@angular/core';
+import {ComponentFixture, TestBed } from '@angular/core/testing';
+import {
+  FlexLayoutModule,
+  MockMatchMedia,
+  MatchMedia,
+  BreakPointsProvider,
+  BreakPointRegistry
+} from '@angular/flex-layout';
+
+describe('layout directive', () => {
+
+  beforeEach(() => {
+
+    // Configure testbed to prepare services
+    TestBed.configureTestingModule({
+      imports: [FlexLayoutModule.forRoot()],
+      declarations: [TestLayoutComponent],
+      providers: [
+        BreakPointRegistry, BreakPointsProvider,
+        {provide: MatchMedia, useClass: MockMatchMedia}
+      ]
+    })
+  });
+
+  it('should add correct styles for default `fxLayout` usage', () => {
+			// TEST LOGIC HERE
+  });
+
+});
+
+/*
+ * Shell component with property 'direction' that will be used
+ * with our tests
+ */
+@Component({
+  selector: 'test-layout',
+  template: `<span>PlaceHolder HTML to be Replaced</span>`
+})
+export class TestLayoutComponent implements OnInit {
+  direction = "column";
+  constructor() {   }
+  ngOnInit() {   }
+}
+
+/*
+ * Custom Helper function to quickly create a `fixture` instance based on
+ * the 'TestLayoutComponent' class
+ */
+function createTestComponent(template: string): ComponentFixture<TestLayoutComponent>
+{
+  return TestBed
+    .overrideComponent(TestLayoutComponent, {set: {template: template}} )
+    .createComponent(TestLayoutComponent);
+}
+{% endraw %}
+{% endhighlight %}
+
+I now have everything I need to write a Directive test quickly.
+
+{% highlight js %}
+{% raw %}
+it('should compile with custom directives', () => {
+  let fixture = createTestComponent(`<div fxLayout></div>`);
+  expect( fixture ).toBeDefined();
+});
+{% endraw %}
+{% endhighlight %}
+
+Wow! That is pretty easy... and the component has been constructed and prepared with the same 
+processes and DI that your real world application would use.
+
+### 2) Using DRY Helper Methods
+
+Let's first write out test using the traditional long-form... one without custom matchers and the 
+more advanced helper methods.
+
+Since the **fxLayout** directive will add custom flexbox CSS to the host element, our test logic 
+will confirm that the initial CSS is correctly assigned.
+
+###### Test Directive Logic: Long-Form
+
+The traditional approach would probably implement something like this:
+
+![pic4](https://cloud.githubusercontent.com/assets/210413/21525027/fd15cf06-ccdf-11e6-8da2-3ac91d5b9d08.png)
```

Comment:
Can you make this a code blog instead? Text scales better than images and can be changed easier in the future when there are API changes.